<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork865-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo"  />
    </a>
</p>


# **SpaceX  Falcon 9 first stage Landing Prediction**


# Lab 1: Collecting the data


Estimated time needed: **45** minutes


In this capstone, we will predict if the Falcon 9 first stage will land successfully. SpaceX advertises Falcon 9 rocket launches on its website with a cost of 62 million dollars; other providers cost upward of 165 million dollars each, much of the savings is because SpaceX can reuse the first stage. Therefore if we can determine if the first stage will land, we can determine the cost of a launch. This information can be used if an alternate company wants to bid against SpaceX for a rocket launch. In this lab, you will collect and make sure the data is in the correct format from an API. The following is an example of a successful and launch.


이번 캡스톤에서는 팰컨9 1단의 성공적 착륙 여부를 예측해보겠습니다. SpaceX는 자사 웹사이트에 6,200만 달러의 비용으로 Falcon 9 로켓 발사를 광고합니다. 다른 제공업체는 각각 1억 6,500만 달러 이상의 비용을 지출하는데, 이러한 절감 효과 중 상당 부분은 SpaceX가 첫 번째 단계를 재사용할 수 있기 때문입니다. 따라서 첫 번째 단계의 착륙 여부를 결정할 수 있다면 발사 비용도 결정할 수 있습니다. 이 정보는 대체 회사가 로켓 발사를 위해 SpaceX에 입찰하려는 경우 사용될 수 있습니다. 이 실습에서는 API에서 데이터를 수집하고 올바른 형식인지 확인합니다. 다음은 성공적인 출시의 예입니다.

![](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/lab_v2/images/landing_1.gif)


Several examples of an unsuccessful landing are shown here:


![](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/lab_v2/images/crash.gif)


Most unsuccessful landings are planned. Space X performs a controlled landing in the oceans. 


대부분의 실패한 착륙이 계획되어 있습니다. Space X는 바다에 통제된 착륙을 수행합니다.

## Objectives


In this lab, you will make a get request to the SpaceX API. You will also do some basic data wrangling and formating. 

- Request to the SpaceX API
- Clean the requested data


----


## Import Libraries and Define Auxiliary Functions


We will import the following libraries into the lab


In [1]:
# Requests allows us to make HTTP requests which we will use to get data from an API
import requests
# Pandas is a software library written for the Python programming language for data manipulation and analysis.
import pandas as pd
# NumPy is a library for the Python programming language, adding support for large, multi-dimensional arrays and matrices, along with a large collection of high-level mathematical functions to operate on these arrays
import numpy as np
# Datetime is a library that allows us to represent dates
import datetime

# Setting this option will print all collumns of a dataframe
pd.set_option('display.max_columns', None)
# Setting this option will print all of the data in a feature
pd.set_option('display.max_colwidth', None)

Below we will define a series of helper functions that will help us use the API to extract information using identification numbers in the launch data.

From the <code>rocket</code> column we would like to learn the booster name.


In [2]:
# Takes the dataset and uses the rocket column to call the API and append the data to the list
def getBoosterVersion(data):
    '''
    Takes the dataset and uses the rocket column to call the API and append the data to the list
    '''
    for x in data['rocket']:
       if x:
        response = requests.get("https://api.spacexdata.com/v4/rockets/"+str(x)).json()
        BoosterVersion.append(response['name'])

From the <code>launchpad</code> we would like to know the name of the launch site being used, the logitude, and the latitude.


In [3]:
# Takes the dataset and uses the launchpad column to call the API and append the data to the list
def getLaunchSite(data):
    '''
    Takes the dataset and uses the launchpad column to call the API and append the data to the list
    '''
    for x in data['launchpad']:
       if x:
         response = requests.get("https://api.spacexdata.com/v4/launchpads/"+str(x)).json()
         Longitude.append(response['longitude'])
         Latitude.append(response['latitude'])
         LaunchSite.append(response['name'])

From the <code>payload</code> we would like to learn the mass of the payload and the orbit that it is going to.


In [4]:
# Takes the dataset and uses the payloads column to call the API and append the data to the lists
def getPayloadData(data):
    '''
    Takes the dataset and uses the payloads column to call the API and append the data to the lists
    '''
    for load in data['payloads']:
       if load:
        response = requests.get("https://api.spacexdata.com/v4/payloads/"+load).json()
        PayloadMass.append(response['mass_kg'])
        Orbit.append(response['orbit'])

From <code>cores</code> we would like to learn the outcome of the landing, the type of the landing, number of flights with that core, whether gridfins were used, wheter the core is reused, wheter legs were used, the landing pad used, the block of the core which is a number used to seperate version of cores, the number of times this specific core has been reused, and the serial of the core.


코어로부터 우리는 착지 결과, 착지 유형, 해당 코어를 사용한 비행 횟수, 그리드핀 사용 여부, 코어 재사용 여부, 다리 사용 여부, 사용된 랜딩 패드, 블록 등을 배우고 싶습니다. 코어의 버전을 구분하는 데 사용되는 숫자인 코어, 이 특정 코어가 재사용된 횟수, 코어의 일련번호입니다.

In [ ]:
# 값이 없는지 : xxx != None

# Null 추가 : xxx.append(None)


In [5]:
# Takes the dataset and uses the cores column to call the API and append the data to the lists
def getCoreData(data):
    '''
    Takes the dataset and uses the cores column to call the API and append the data to the lists
    '''
    for core in data['cores']:
            if core['core'] != None:
                response = requests.get("https://api.spacexdata.com/v4/cores/"+core['core']).json()
                Block.append(response['block'])
                ReusedCount.append(response['reuse_count'])
                Serial.append(response['serial'])
            else:
                Block.append(None)
                ReusedCount.append(None)
                Serial.append(None)
            Outcome.append(str(core['landing_success'])+' '+str(core['landing_type']))
            Flights.append(core['flight'])
            GridFins.append(core['gridfins'])
            Reused.append(core['reused'])
            Legs.append(core['legs'])
            LandingPad.append(core['landpad'])

Now let's start requesting rocket launch data from SpaceX API with the following URL:


In [6]:
spacex_url="https://api.spacexdata.com/v4/launches/past"

In [7]:
response = requests.get(spacex_url)

Check the content of the response


In [8]:
# print(response.content)

# 앞 xxx 글자만 출력
print( response.content [:900] )

b'[{"fairings":{"reused":false,"recovery_attempt":false,"recovered":false,"ships":[]},"links":{"patch":{"small":"https://images2.imgbox.com/94/f2/NN6Ph45r_o.png","large":"https://images2.imgbox.com/5b/02/QcxHUb5V_o.png"},"reddit":{"campaign":null,"launch":null,"media":null,"recovery":null},"flickr":{"small":[],"original":[]},"presskit":null,"webcast":"https://www.youtube.com/watch?v=0a_00nJ_Y88","youtube_id":"0a_00nJ_Y88","article":"https://www.space.com/2196-spacex-inaugural-falcon-1-rocket-lost-launch.html","wikipedia":"https://en.wikipedia.org/wiki/DemoSat"},"static_fire_date_utc":"2006-03-17T00:00:00.000Z","static_fire_date_unix":1142553600,"net":false,"window":0,"rocket":"5e9d0d95eda69955f709d1eb","success":false,"failures":[{"time":33,"altitude":null,"reason":"merlin engine failure"}],"details":"Engine failure at 33 seconds and loss of vehicle","crew":[],"ships":[],"capsules":[],"pay'


In [9]:
# test
len( response.content )

416346

You should see the response contains massive information about SpaceX launches. Next, let's try to discover some more relevant information for this project.


### Task 1: Request and parse the SpaceX launch data using the GET request


To make the requested JSON results more consistent, we will use the following static response object for this project:


In [10]:
static_json_url='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/API_call_spacex_api.json'

We should see that the request was successfull with the 200 status response code


In [11]:
response.status_code

200

Now we decode the response content as a Json using <code>.json()</code> and turn it into a Pandas dataframe using <code>.json_normalize()</code>


이제 .json()을 사용하여 응답 콘텐츠를 Json으로 디코딩하고 .json_normalize()를 사용하여 Pandas 데이터 프레임으로 변환합니다.

In [18]:
# Use json_normalize meethod to convert the json result into a dataframe

data = response.json().json_normalize()

AttributeError: 'list' object has no attribute 'json_normalize'

In [12]:
# 설명도 안 해 주고 하라니. 헐.

# json 디코딩
response_json = response.json()
data = pd.json_normalize( response_json )

Using the dataframe <code>data</code> print the first 5 rows


In [13]:
# Get the head of the dataframe

data.head()

static_fire_date_utc  static_fire_date_unix    net  window  \
0  2006-03-17T00:00:00.000Z           1.142554e+09  False     0.0   
1                      None                    NaN  False     0.0   
2                      None                    NaN  False     0.0   
3  2008-09-20T00:00:00.000Z           1.221869e+09  False     0.0   
4                      None                    NaN  False     0.0   

                     rocket success  \
0  5e9d0d95eda69955f709d1eb   False   
1  5e9d0d95eda69955f709d1eb   False   
2  5e9d0d95eda69955f709d1eb   False   
3  5e9d0d95eda69955f709d1eb    True   
4  5e9d0d95eda69955f709d1eb    True   

                                                                                                            failures  \
0                                                [{'time': 33, 'altitude': None, 'reason': 'merlin engine failure'}]   
1            [{'time': 301, 'altitude': 289, 'reason': 'harmonic oscillation leading to premature engine shutdown'}]   
2  [{'time': 140, 'altitude': 35, 'reason': 'residual stage-1 thrust led to collision between stage 1 and stage 2'}]   
3                                                                                                                 []   
4                                                                                                                 []   

                                                                                                                                                                                details  \
0                                                                                                                                      Engine failure at 33 seconds and loss of vehicle   
1  Successful first stage burn and transition to second stage, maximum altitude 289 km, Premature engine shutdown at T+7 min 30 s, Failed to reach orbit, Failed to recover first stage   
2                                                                                                                  Residual stage 1 thrust led to collision between stage 1 and stage 2   
3                        Ratsat was carried to orbit on the first successful orbital launch of any privately funded and developed, liquid-propelled carrier rocket, the SpaceX Falcon 1   
4                                                                                                                                                                                  None   

  crew ships capsules                                              payloads  \
0   []    []       []                            [5eb0e4b5b6c3bb0006eeb1e1]   
1   []    []       []                            [5eb0e4b6b6c3bb0006eeb1e2]   
2   []    []       []  [5eb0e4b6b6c3bb0006eeb1e3, 5eb0e4b6b6c3bb0006eeb1e4]   
3   []    []       []                            [5eb0e4b7b6c3bb0006eeb1e5]   
4   []    []       []                            [5eb0e4b7b6c3bb0006eeb1e6]   

                  launchpad  flight_number         name  \
0  5e9e4502f5090995de566f86              1    FalconSat   
1  5e9e4502f5090995de566f86              2      DemoSat   
2  5e9e4502f5090995de566f86              3  Trailblazer   
3  5e9e4502f5090995de566f86              4       RatSat   
4  5e9e4502f5090995de566f86              5     RazakSat   

                   date_utc   date_unix                 date_local  \
0  2006-03-24T22:30:00.000Z  1143239400  2006-03-25T10:30:00+12:00   
1  2007-03-21T01:10:00.000Z  1174439400  2007-03-21T13:10:00+12:00   
2  2008-08-03T03:34:00.000Z  1217734440  2008-08-03T15:34:00+12:00   
3  2008-09-28T23:15:00.000Z  1222643700  2008-09-28T11:15:00+12:00   
4  2009-07-13T03:35:00.000Z  1247456100  2009-07-13T15:35:00+12:00   

  date_precision  upcoming  \
0           hour     False   
1           hour     False   
2           hour     False   
3           hour     False   
4           hour     False   

                                                                                 

In [14]:
# test
data.shape

(187, 43)

You will notice that a lot of the data are IDs. For example the rocket column has no information about the rocket just an identification number.

We will now use the API again to get information about the launches using the IDs given for each launch. Specifically we will be using columns <code>rocket</code>, <code>payloads</code>, <code>launchpad</code>, and <code>cores</code>.


In [ ]:
# 배울 만한 구문들이 있는 듯.

# 통으로 되어 있어서, 좀 split 해야겠네.

G

이 코드는 주어진 데이터 프레임 `data`에서 특정 열을 선택하고 데이터를 필터링, 가공하는 작업을 수행합니다. 주요 단계와 설명은 다음과 같습니다:

1. 먼저, 주요한 열인 `'rocket'`, `'payloads'`, `'launchpad'`, `'cores'`, `'flight_number'`, 그리고 `'date_utc'`만 선택하여 데이터 프레임을 갱신합니다.

2. 다음으로, `'cores'`와 `'payloads'` 열에서 리스트 형태로 된 값들을 가진 행을 필터링합니다. 이렇게 하는 이유는 다중 코어를 가진 로켓과 다중 페이로드를 가진 로켓을 제외하기 위함입니다.

3. 다중 코어 및 다중 페이로드를 제외한 나머지 행들에 대해 `'cores'`와 `'payloads'` 열의 값을 리스트에서 첫 번째 요소만 추출하여 대체합니다.

4. `'date_utc'` 열의 값을 날짜와 시간 형식에서 날짜만을 추출하고 이를 `'date'` 열에 저장합니다.

5. 마지막으로, `'date'` 열을 기준으로 날짜가 '2020-11-13' 이하인 행들만을 남기고, 이를 다시 데이터 프레임에 할당하여 데이터를 제한합니다.

이러한 작업을 통해 원하는 열만 남기고, 데이터를 필터링하고, 날짜를 가공하여 최종적으로 원하는 형태의 데이터 프레임을 얻을 수 있게 됩니다.

In [16]:
# Lets take a subset of our dataframe keeping only the features we want and the flight number, 
# and date_utc.
data = data[['rocket', 'payloads', 'launchpad', 'cores', 'flight_number', 'date_utc']]

In [19]:
# We will remove rows with multiple cores 
# because those are falcon rockets with 2 extra rocket boosters and rows 
# that have multiple payloads in a single rocket.
data = data[data['cores'].map(len)==1]
data = data[data['payloads'].map(len)==1]
# map(적용 할 함수) : 각 list의 값에 대해 파라미터의 함수를 적용함.
# data['cores'].map(len)==1 : 즉, list가 값 1개만 있는 것만 True 리턴

In [23]:
# test
data['cores'].map(len)==1

0      True
1      True
3      True
4      True
5      True
       ... 
180    True
181    True
184    True
185    True
186    True
Name: cores, Length: 172, dtype: bool

In [22]:
# test
data[data['cores'].map(len)==1].head(2)

rocket                    payloads  \
0  5e9d0d95eda69955f709d1eb  [5eb0e4b5b6c3bb0006eeb1e1]   
1  5e9d0d95eda69955f709d1eb  [5eb0e4b6b6c3bb0006eeb1e2]   

                  launchpad  \
0  5e9e4502f5090995de566f86   
1  5e9e4502f5090995de566f86   

                                                                                                                                                                                              cores  \
0  [{'core': '5e9e289df35918033d3b2623', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}]   
1  [{'core': '5e9e289ef35918416a3b2624', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}]   

   flight_number                  date_utc  
0              1  2006-03-24T22:30:00.000Z  
1              2  2007-03-21T01:10:00.000Z

In [24]:
# Since payloads and cores are lists of size 1 
# we will also extract the single value in the list and replace the feature.
data['cores'] = data['cores'].map(lambda x : x[0])
data['payloads'] = data['payloads'].map(lambda x : x[0])
# df 해당 컬럼에 대해, 각 값을 map 안의 lambda(익명 함수)적용
# 익명 함수의 파라미터 x에 각 값을 넣은 후, x[0] 리턴.

In [25]:
# test
data.head(2)

rocket                  payloads  \
0  5e9d0d95eda69955f709d1eb  5eb0e4b5b6c3bb0006eeb1e1   
1  5e9d0d95eda69955f709d1eb  5eb0e4b6b6c3bb0006eeb1e2   

                  launchpad  \
0  5e9e4502f5090995de566f86   
1  5e9e4502f5090995de566f86   

                                                                                                                                                                                            cores  \
0  {'core': '5e9e289df35918033d3b2623', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}   
1  {'core': '5e9e289ef35918416a3b2624', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}   

   flight_number                  date_utc  
0              1  2006-03-24T22:30:00.000Z  
1              2  2007-03-21T01:10:00.000Z

G

`lambda`는 파이썬에서 익명 함수(anonymous function)를 생성하는데 사용되는 키워드입니다. 익명 함수란 이름이 없는, 일회성으로 사용되는 함수를 의미합니다. `lambda` 표현식은 간단한 함수를 간결하게 정의하고 인라인으로 사용할 때 유용합니다.

`lambda` 함수의 기본 구문은 다음과 같습니다:

```python
lambda arguments: expression
```

- `arguments`: 함수의 인자(매개변수)를 나타내며, 콜론(`:`)으로 구분됩니다.
- `expression`: 인자를 이용하여 계산하고 반환할 표현식을 나타냅니다.

`lambda` 함수는 주로 다른 함수나 메서드의 인자로 전달되거나, 간단한 연산을 필요로 하는 곳에서 사용됩니다. 특히, `map()` 함수나 `filter()` 함수와 함께 사용하는 경우가 많습니다.

예를 들어, 리스트의 각 요소에 대해 제곱한 결과를 얻고 싶다면 다음과 같이 `lambda` 함수를 사용할 수 있습니다:

```python
numbers = [1, 2, 3, 4, 5]
squared_numbers = list(map(lambda x: x**2, numbers))
print(squared_numbers)  # 출력: [1, 4, 9, 16, 25]
```

위 코드에서 `lambda x: x**2`는 인자 `x`를 받아서 `x`의 제곱을 반환하는 익명 함수를 나타냅니다.

따라서 `data['cores'] = data['cores'].map(lambda x : x[0])` 코드는 `data['cores']` 시리즈의 각 요소(`x`)에 대해 첫 번째 원소(`x[0]`)를 추출하는 `lambda` 함수를 사용하여 시리즈의 값을 변환합니다. 이렇게 하면 `cores` 시리즈의 값이 리스트 형태에서 첫 번째 원소만 포함된 값으로 변환됩니다.

In [26]:
# We also want to convert the date_utc to a datetime datatype 
# aand then extracting the date leaving the time
data['date'] = pd.to_datetime(data['date_utc']).dt.date

In [27]:
# test
data.head(2)

rocket                  payloads  \
0  5e9d0d95eda69955f709d1eb  5eb0e4b5b6c3bb0006eeb1e1   
1  5e9d0d95eda69955f709d1eb  5eb0e4b6b6c3bb0006eeb1e2   

                  launchpad  \
0  5e9e4502f5090995de566f86   
1  5e9e4502f5090995de566f86   

                                                                                                                                                                                            cores  \
0  {'core': '5e9e289df35918033d3b2623', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}   
1  {'core': '5e9e289ef35918416a3b2624', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}   

   flight_number                  date_utc        date  
0              1  2006-03-24T22:30:00.000Z  2006-03-24  
1              2  2007-03-21T01:10:00.000Z  2007-03-21

G

위 파이썬 코드는 `date_utc` 열의 값을 datetime 형식으로 변환하고, 그 날짜에서 시간 정보를 제외하여 `date` 열에 저장하는 작업을 수행합니다.

1. `pd.to_datetime(data['date_utc'])`: `date_utc` 열의 값을 datetime 형식으로 변환합니다. 즉, 날짜와 시간 정보가 포함된 datetime 객체로 변환됩니다.

2. `.dt.date`: datetime 객체에서 날짜 정보만 추출하여 새로운 열인 `date` 열에 저장합니다. 따라서 `date` 열에는 날짜 정보만 있는 것이고, 시간 정보는 제외됩니다.

이렇게 하면 `date_utc` 열의 값들이 날짜와 시간 정보를 포함하는 문자열이었던 것을, 날짜 정보만을 가지는 `date` 열로 변환할 수 있습니다. 이렇게 변환된 데이터는 날짜를 기준으로 데이터를 필터링하거나 시간 관련 작업을 수행하는 데 유용할 수 있습니다.

In [28]:
# Using the date we will restrict the dates of the launches
data = data[data['date'] <= datetime.date(2020, 11, 13)]

In [30]:
# test
data.tail(2)

rocket                  payloads  \
104  5e9d0d95eda69973a809d1ec  5ef6a4ea0059c33cee4a82a2   
105  5e9d0d95eda69973a809d1ec  5eb0e4d2b6c3bb0006eeb25e   

                    launchpad  \
104  5e9e4501f509094ba4566f84   
105  5e9e4501f509094ba4566f84   

                                                                                                                                                                                                                   cores  \
104   {'core': '5ef670f10059c33cee4a826c', 'flight': 3, 'gridfins': True, 'legs': True, 'reused': True, 'landing_attempt': True, 'landing_success': True, 'landing_type': 'ASDS', 'landpad': '5e9e3033383ecbb9e534e7cc'}   
105  {'core': '5f57c5440622a633027900a0', 'flight': 1, 'gridfins': True, 'legs': True, 'reused': False, 'landing_attempt': True, 'landing_success': True, 'landing_type': 'ASDS', 'landpad': '5e9e3032383ecb6bb234e7ca'}   

     flight_number                  date_utc        date  
104            105  2020-10-24T15:31:00.000Z  2020-10-24  
105            106  2020-11-05T23:24:00.000Z  2020-11-05

In [32]:
# test
data.shape

(94, 7)

* From the <code>rocket</code> we would like to learn the booster name

* From the <code>payload</code> we would like to learn the mass of the payload and the orbit that it is going to

* From the <code>launchpad</code> we would like to know the name of the launch site being used, the longitude, and the latitude.

* From <code>cores</code> we would like to learn the outcome of the landing, the type of the landing, number of flights with that core, whether gridfins were used, whether the core is reused, whether legs were used, the landing pad used, the block of the core which is a number used to seperate version of cores, the number of times this specific core has been reused, and the serial of the core.

The data from these requests will be stored in lists and will be used to create a new dataframe.


- 로켓에서 부스터 이름을 알고 싶습니다.

- 페이로드로부터 페이로드의 질량과 페이로드가 이동할 궤도를 배우고 싶습니다.

- 발사대에서 우리는 사용 중인 발사 장소의 이름, 경도, 위도를 알고 싶습니다.

- 코어로부터 착지 결과, 착지 유형, 해당 코어를 사용한 비행 횟수, 그리드핀 사용 여부, 코어 재사용 여부, 다리 사용 여부, 랜딩 패드 사용, 블록 등을 배우고 싶습니다. 코어의 버전을 구분하는 데 사용되는 숫자인 코어, 이 특정 코어가 재사용된 횟수, 코어의 일련번호입니다.

이러한 요청의 데이터는 목록에 저장되며 새 데이터 프레임을 만드는 데 사용됩니다.

In [33]:
#Global variables 
BoosterVersion = []
PayloadMass = []
Orbit = []
LaunchSite = []
Outcome = []
Flights = []
GridFins = []
Reused = []
Legs = []
LandingPad = []
Block = []
ReusedCount = []
Serial = []
Longitude = []
Latitude = []

These functions will apply the outputs globally to the above variables. Let's take a looks at <code>BoosterVersion</code> variable. Before we apply  <code>getBoosterVersion</code> the list is empty:


In [34]:
BoosterVersion

[]

In [38]:
# RestAPI 호출하는데 시간이 많이 걸리니, 수행 시간 체크 해 보자.
import time

Now, let's apply <code> getBoosterVersion</code> function method to get the booster version


In [35]:
# Call getBoosterVersion
getBoosterVersion(data)

the list has now been update 


In [36]:
%time
BoosterVersion[0:5]

CPU times: total: 0 ns
Wall time: 0 ns


['Falcon 1', 'Falcon 1', 'Falcon 1', 'Falcon 1', 'Falcon 9']

we can apply the rest of the  functions here:


In [37]:
# Call getLaunchSite
%time
getLaunchSite(data)

CPU times: total: 0 ns
Wall time: 0 ns


In [39]:
start_time = time.time()

# Call getPayloadData
getPayloadData(data)

end_time = time.time()
execution_time = end_time - start_time
print(f'execution_time : {execution_time} sec')

execution_time : 55.19531488418579 sec


In [40]:
start_time = time.time()

# Call getCoreData
getCoreData(data)

end_time = time.time()
execution_time = end_time - start_time
print(f'execution_time : {execution_time} sec')

execution_time : 54.409555435180664 sec


Finally lets construct our dataset using the data we have obtained. We we combine the columns into a dictionary.


In [41]:
launch_dict = {'FlightNumber': list(data['flight_number']),
               'Date': list(data['date']),
               'BoosterVersion':BoosterVersion,
               'PayloadMass':PayloadMass,
               'Orbit':Orbit,
               'LaunchSite':LaunchSite,
               'Outcome':Outcome,
               'Flights':Flights,
               'GridFins':GridFins,
               'Reused':Reused,
               'Legs':Legs,
               'LandingPad':LandingPad,
               'Block':Block,
               'ReusedCount':ReusedCount,
               'Serial':Serial,
               'Longitude': Longitude,
               'Latitude': Latitude}


Then, we need to create a Pandas data frame from the dictionary launch_dict.


In [42]:
# Create a data from launch_dict

launch_df = pd.DataFrame( launch_dict )

Show the summary of the dataframe


In [66]:
# Show the head of the dataframe

launch_df.head(3)

FlightNumber        Date BoosterVersion  PayloadMass Orbit  \
0             1  2006-03-24       Falcon 1         20.0   LEO   
1             2  2007-03-21       Falcon 1          NaN   LEO   
2             4  2008-09-28       Falcon 1        165.0   LEO   

        LaunchSite    Outcome  Flights  GridFins  Reused   Legs LandingPad  \
0  Kwajalein Atoll  None None        1     False   False  False       None   
1  Kwajalein Atoll  None None        1     False   False  False       None   
2  Kwajalein Atoll  None None        1     False   False  False       None   

   Block  ReusedCount    Serial   Longitude  Latitude  
0    NaN            0  Merlin1A  167.743129  9.047721  
1    NaN            0  Merlin2A  167.743129  9.047721  
2    NaN            0  Merlin2C  167.743129  9.047721

In [67]:
# test
launch_df.shape

(94, 17)

### Task 2: Filter the dataframe to only include `Falcon 9` launches


Finally we will remove the Falcon 1 launches keeping only the Falcon 9 launches. Filter the data dataframe using the <code>BoosterVersion</code> column to only keep the Falcon 9 launches. Save the filtered data to a new dataframe called <code>data_falcon9</code>.


In [68]:
# Hint data['BoosterVersion']!='Falcon 1'

data_falcon9 = launch_df[ launch_df['BoosterVersion'] != 'Falcon 1' ]
data_falcon9.head(3)

FlightNumber        Date BoosterVersion  PayloadMass Orbit    LaunchSite  \
4             6  2010-06-04       Falcon 9          NaN   LEO  CCSFS SLC 40   
5             8  2012-05-22       Falcon 9        525.0   LEO  CCSFS SLC 40   
6            10  2013-03-01       Falcon 9        677.0   ISS  CCSFS SLC 40   

     Outcome  Flights  GridFins  Reused   Legs LandingPad  Block  ReusedCount  \
4  None None        1     False   False  False       None    1.0            0   
5  None None        1     False   False  False       None    1.0            0   
6  None None        1     False   False  False       None    1.0            0   

  Serial  Longitude   Latitude  
4  B0003 -80.577366  28.561857  
5  B0005 -80.577366  28.561857  
6  B0007 -80.577366  28.561857

In [44]:
# test
launch_df['BoosterVersion'] != 'Falcon 1'

0     False
1     False
2     False
3     False
4      True
      ...  
89     True
90     True
91     True
92     True
93     True
Name: BoosterVersion, Length: 94, dtype: bool

In [46]:
# test
launch_df.groupby(['BoosterVersion']) ['BoosterVersion'].count()

BoosterVersion
Falcon 1     4
Falcon 9    90
Name: BoosterVersion, dtype: int64

Now that we have removed some values we should reset the FlgihtNumber column


이제 일부 값을 제거했으므로 FlgihtNumber 열을 재설정해야 합니다.
> 왜???
>> FlightNumber 값을 seq하게 변경하기 보다는 index를 reset 하는 것이 낫지 않을까?

In [69]:
data_falcon9.loc[:,'FlightNumber'] = list(range(1, data_falcon9.shape[0]+1))
data_falcon9

C:\Users\jsh\AppData\Local\Temp\ipykernel_11284\4023201451.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_falcon9.loc[:,'FlightNumber'] = list(range(1, data_falcon9.shape[0]+1))


FlightNumber        Date BoosterVersion  PayloadMass Orbit    LaunchSite  \
4              1  2010-06-04       Falcon 9          NaN   LEO  CCSFS SLC 40   
5              2  2012-05-22       Falcon 9        525.0   LEO  CCSFS SLC 40   
6              3  2013-03-01       Falcon 9        677.0   ISS  CCSFS SLC 40   
7              4  2013-09-29       Falcon 9        500.0    PO   VAFB SLC 4E   
8              5  2013-12-03       Falcon 9       3170.0   GTO  CCSFS SLC 40   
..           ...         ...            ...          ...   ...           ...   
89            86  2020-09-03       Falcon 9      15600.0  VLEO    KSC LC 39A   
90            87  2020-10-06       Falcon 9      15600.0  VLEO    KSC LC 39A   
91            88  2020-10-18       Falcon 9      15600.0  VLEO    KSC LC 39A   
92            89  2020-10-24       Falcon 9      15600.0  VLEO  CCSFS SLC 40   
93            90  2020-11-05       Falcon 9       3681.0   MEO  CCSFS SLC 40   

        Outcome  Flights  GridFins  Reused   Legs                LandingPad  \
4     None None        1     False   False  False                      None   
5     None None        1     False   False  False                      None   
6     None None        1     False   False  False                      None   
7   False Ocean        1     False   False  False                      None   
8     None None        1     False   False  False                      None   
..          ...      ...       ...     ...    ...                       ...   
89    True ASDS        2      True    True   True  5e9e3032383ecb6bb234e7ca   
90    True ASDS        3      True    True   True  5e9e3032383ecb6bb234e7ca   
91    True ASDS        6      True    True   True  5e9e3032383ecb6bb234e7ca   
92    True ASDS        3      True    True   True  5e9e3033383ecbb9e534e7cc   
93    True ASDS        1      True   False   True  5e9e3032383ecb6bb234e7ca   

    Block  ReusedCount Serial   Longitude   Latitude  
4     1.0            0  B0003  -80.577366  28.561857  
5     1.0            0  B0005  -80.577366  28.561857  
6     1.0            0  B0007  -80.577366  28.561857  
7     1.0            0  B1003 -120.610829  34.632093  
8     1.0            0  B1004  -80.577366  28.561857  
..    ...          ...    ...         ...        ...  
89    5.0           12  B1060  -80.603956  28.608058  
90    5.0           13  B1058  -80.603956  28.608058  
91    5.0           12  B1051  -80.603956  28.608058  
92    5.0           12  B1060  -80.577366  28.561857  
93    5.0            8  B1062  -80.577366  28.561857  

[90 rows x 17 columns]

In [70]:
# test
data_falcon9.shape

(90, 17)

## Data Wrangling


We can see below that some of the rows are missing values in our dataset.


In [ ]:
# 이것도 좋군.

In [71]:
data_falcon9.isnull().sum()

FlightNumber       0
Date               0
BoosterVersion     0
PayloadMass        5
Orbit              0
LaunchSite         0
Outcome            0
Flights            0
GridFins           0
Reused             0
Legs               0
LandingPad        26
Block              0
ReusedCount        0
Serial             0
Longitude          0
Latitude           0
dtype: int64

Before we can continue we must deal with these missing values. The <code>LandingPad</code> column will retain None values to represent when landing pads were not used.


### Task 3: Dealing with Missing Values


Calculate below the mean for the <code>PayloadMass</code> using the <code>.mean()</code>. Then use the mean and the <code>.replace()</code> function to replace `np.nan` values in the data with the mean you calculated.


In [72]:
# Calculate the mean value of PayloadMass column

PayloadMass_mean = data_falcon9['PayloadMass'].mean()
PayloadMass_mean

6123.547647058824

In [76]:
# Replace the np.nan values with its mean value

# 모든 컬럼에 대해서 하므로, 틀림.
# data_falcon9.fillna(PayloadMass_mean, inplace=True)

# df 복사
data_falcon9_nn = data_falcon9.copy(deep=True)
# nn : not null
data_falcon9_nn['PayloadMass'] = data_falcon9['PayloadMass'].fillna(PayloadMass_mean)

In [77]:
# test
data_falcon9_nn.isnull().sum()

FlightNumber       0
Date               0
BoosterVersion     0
PayloadMass        0
Orbit              0
LaunchSite         0
Outcome            0
Flights            0
GridFins           0
Reused             0
Legs               0
LandingPad        26
Block              0
ReusedCount        0
Serial             0
Longitude          0
Latitude           0
dtype: int64

In [78]:
data_falcon9_nn.head(3)

FlightNumber        Date BoosterVersion  PayloadMass Orbit    LaunchSite  \
4             1  2010-06-04       Falcon 9  6123.547647   LEO  CCSFS SLC 40   
5             2  2012-05-22       Falcon 9   525.000000   LEO  CCSFS SLC 40   
6             3  2013-03-01       Falcon 9   677.000000   ISS  CCSFS SLC 40   

     Outcome  Flights  GridFins  Reused   Legs LandingPad  Block  ReusedCount  \
4  None None        1     False   False  False       None    1.0            0   
5  None None        1     False   False  False       None    1.0            0   
6  None None        1     False   False  False       None    1.0            0   

  Serial  Longitude   Latitude  
4  B0003 -80.577366  28.561857  
5  B0005 -80.577366  28.561857  
6  B0007 -80.577366  28.561857

In [ ]:
# 아래는, null 제거 전.

In [54]:
# test
data_falcon9[ data_falcon9['PayloadMass'] == np.nan ]  # 조회 0건

Empty DataFrame
Columns: [FlightNumber, Date, BoosterVersion, PayloadMass, Orbit, LaunchSite, Outcome, Flights, GridFins, Reused, Legs, LandingPad, Block, ReusedCount, Serial, Longitude, Latitude]
Index: []

In [58]:
# test
data_falcon9[ data_falcon9['PayloadMass'].isnull() ]

FlightNumber        Date BoosterVersion  PayloadMass Orbit    LaunchSite  \
4              1  2010-06-04       Falcon 9          NaN   LEO  CCSFS SLC 40   
33            30  2017-05-01       Falcon 9          NaN   LEO    KSC LC 39A   
47            44  2018-01-08       Falcon 9          NaN   LEO  CCSFS SLC 40   
76            73  2020-01-19       Falcon 9          NaN    SO    KSC LC 39A   
86            83  2020-07-20       Falcon 9          NaN   GEO  CCSFS SLC 40   

      Outcome  Flights  GridFins  Reused   Legs                LandingPad  \
4   None None        1     False   False  False                      None   
33  True RTLS        1      True   False   True  5e9e3032383ecb267a34e7c7   
47  True RTLS        1      True   False   True  5e9e3032383ecb267a34e7c7   
76  None None        4     False    True  False                      None   
86  True ASDS        2      True    True   True  5e9e3033383ecbb9e534e7cc   

    Block  ReusedCount Serial  Longitude   Latitude  
4     1.0            0  B0003 -80.577366  28.561857  
33    3.0            1  B1032 -80.603956  28.608058  
47    4.0            1  B1043 -80.577366  28.561857  
76    5.0            3  B1046 -80.603956  28.608058  
86    5.0           13  B1058 -80.577366  28.561857

You should see the number of missing values of the <code>PayLoadMass</code> change to zero.


Now we should have no missing values in our dataset except for in <code>LandingPad</code>.


We can now export it to a <b>CSV</b> for the next section,but to make the answers consistent, in the next lab we will provide data in a pre-selected date range. 


<code>data_falcon9.to_csv('dataset_part_1.csv', index=False)</code>


In [79]:
data_falcon9_nn.to_csv('dataset_part_1.csv', index=False)

## Authors


<a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork865-2022-01-01">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD. 


## Change Log


|Date (YYYY-MM-DD)|Version|Changed By|Change Description|
|-|-|-|-|
|2020-09-20|1.1|Joseph|get result each time you run|
|2020-09-20|1.1|Azim |Created Part 1 Lab using SpaceX API|
|2020-09-20|1.0|Joseph |Modified Multiple Areas|


Copyright © 2021 IBM Corporation. All rights reserved.
